File: tink-cc-wrapped/src/cleartext_keyset_handle.cpp
#include "cleartext_keyset_handle.h"

OwnedTinkStatusOrKeysetHandle wrapped_CleartextKeysetHandle_Read(OwnedKeysetReader reader) {
    std::unique_ptr<crypto::tink::KeysetReader> cc_reader(reader.cc_keyset_reader);
    absl::StatusOr<std::unique_ptr<crypto::tink::KeysetHandle>> result = crypto::tink::CleartextKeysetHandle::Read(std::move(cc_reader));
    return toOwnedTinkStatusOrKeysetHandle(std::move(result));
}

File: src/cleartext_keyset_handle.rs
use crate::keyset_handle_transfer::OwnedTinkStatusOrKeysetHandle;
use crate::keyset_reader_transfer::OwnedKeysetReader;
use crate::KeysetHandle;
use crate::KeysetReader;
use crate::TinkError;

use crate::cleartext_keyset_handle_transfer::*;

// --------------------------------------------------------------
//
// Public Rust API
//

/// Creates keyset handles from cleartext keysets. This API allows
/// loading cleartext keysets, thus its usage should be restricted.
pub struct CleartextKeysetHandle {}
impl CleartextKeysetHandle {
    /// Creates a KeysetHandle with a keyset obtained via `reader`. Optionally
    /// allows to pass monitoring_annotations to attach additional data to the
    /// resulting KeysetHandle, which will be used for monitoring.
    pub fn read(reader: KeysetReader) -> Result<KeysetHandle, TinkError> {
        unsafe {
            let xfer_reader: OwnedKeysetReader = reader.into();
            let result: OwnedTinkStatusOrKeysetHandle =
                wrapped_CleartextKeysetHandle_Read(xfer_reader);
            result.into()
        }
    }
}
